### ISSUE TITLE
[Doc] Python API KV/memory reset details absent

### ISSUE THREAD

**Issue by federicoparra** [mlc-ai/mlc-llm#2426]
## 📚 Documentation

Hi, I just migrated to last version of MLC with new MLCengine from the old chat engine. In the old one the prompts and inferences were concatenated automatically by default and one had to reset it to stop it from doing so with the reset() method.

In the new create() method which imitates openAI API the method is stateless so it's our responsibility to keep a list of prompts and inferences and send them each time we call create() in order for the LLM to have memory.

However, it appears that despite this, the KV cache of the model does seem to be kept from call to call; I'm guessing this because when called with a long list of messages it can very quickly begin the next inference, as if it would just use the previous KV cache from the previous messages/inferences. 

All of this should be very explicit in the documentation, including if there is a command (like the old reset) to fully reset the KV cache.

Also, it would be nice to be able to send the engine a KV cache (and for it to return one) - for example when we have complex and long system prompts, we could run it once and keep the kv cache and then when we call the LLM again for inference on that prompt we would send the KV cache for that prompt and save a lot of time correct? We can do this with the Transformers API for instance.

Thank you for all the hard work!!!!
Federico

**Comment by tqchen** [comment]
Thanks for the suggestion. https://github.com/mlc-ai/mlc-llm/blob/main/python/mlc_llm/interface/chat.py#L173 contains some examples of chat session.

We did have kv cache cahing internally via recent #2295 , that means you can just send the same requests (with system prompts) and it will automatically cache as long as the same prefix is cached.

**Comment by federicoparra** [comment]
Here's an example: say you want to use an LLM as a form of general intelligence that helps perform several specific functions such as:

1) extract specific words from a text
2) summarize a text
3) translate a text
etc

The prompts required for each of those functions can be somewhat long if it contains details about the task. The time it takes for the model to process the prompts (to encode them) each time we want to execute one of these functions is unnecessary - with transformers API we would run it once, save the KV cache to a file, then load the data every time we want to perform that function.

**Comment by tqchen** [comment]
Indeed you are right that it is important to be able to handle cached prompt prefix so we don't have to recompute these kv again. In our case, let us say the requests comes with

- R0: <translate a text prompt>[user specific request0]
- R1: <translate a text prompt> [user specific request1]
- R2: <translate a text prompt> [user specific request2]


When R0 hit the system, we will run that long encoding, then when R1 and R2 hit the system, `<translate a text prompt> ` would be cached in system automatically so effectively we only run compute for `[user specific request1]` and `[user specific request2]`

Say in a more complicated case

- R0: <translate a text prompt>[user specific request0]
- R1: <translate a text prompt> [user specific request1]
- R2: <summarize a text prompt> [user specific request2]
- R3: <translate a text prompt> [user specific request3]
- R4: <summarize a text prompt> [user specific request4]

In this case, depending on how many prefix the system decides to keep around, likely for all request only user specific requests are handled. To guarantee such effect, we can also introduce a pin system prompt special request to allow user to specifically pin those prompts, and we guarantee won't recyling them. So these prompts are processed once during the entire interactions

**Comment by federicoparra** [comment]
I like your idea! 

I also like the straightforward way Transformers library does it: their Generate method simply allows us to provide kvs as an input, and also offers then as an output. 

That allows for maximum flexibility: one can even save them to hard drive and create a library of preprocessed prompts to then feed the model for very quick inference.

**Comment by tqchen** [comment]
Indeed, having ability to feed in kv is useful. Because right now we design for concurrent access(from multiple users), so doing so in general is harder. We will also think about how to design an interface to allow feeding

**Comment by federicoparra** [comment]
In general though, what we need is more documentation (for those that are lazy or don't know how to look at all the code like me :) about the state handling in the new API (MLCengine); for example if I am deducing from what you are saying it looks like sometimes there is pinning of KVs if they are sent as system prompts, but not always; it also looks like by default the engine is stateless (i.e., if we only send user prompts and assistant responses, they are not in any way cached) but when I try I notice that there is a speed up once the API receives a user/assistant prompt it has already seen so I suspect there is some caching behind the scenes - this is all a little obscure. I would argue it would be good to have a few paragraphs about this, maybe in a specific documentation page about state handling. 

The idea of having system prompts be pinned is amazing by the way. That way whatever the window size is, the most important piece of context (the instructions in the system message) are always remembered!
